### ISSUE TITLE
slow typeclass synthesis: takes 19000 heartbeats to fail on `AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ`

### ISSUE THREAD

**Issue by semorrison** [leanprover-community/mathlib4#12226]
[live.lean-lang.org](https://live.lean-lang.org/#code=import%20Mathlib%0A%0Aopen%20Complex%20Filter%20Bornology%0A%0A%2F--%0Aerror%3A%20failed%20to%20synthesize%0A%20%20AddMonoidHomClass%20(AddGroupSeminorm%20%E2%84%82)%20%E2%84%82%20%E2%84%9D%0A(deterministic)%20timeout%20at%20'typeclass'%2C%20maximum%20number%20of%20heartbeats%20(19000)%20has%20been%20reached%20(use%20'set_option%20synthInstance.maxHeartbeats%20%3Cnum%3E'%20to%20set%20the%20limit)%0A-%2F%0A%23guard_msgs%20in%0Aset_option%20synthInstance.maxHeartbeats%2019000%20in%0A%23synth%20AddMonoidHomClass%20(AddGroupSeminorm%20%E2%84%82)%20%E2%84%82%20%E2%84%9D%0A%0A--%20This%20then%20results%20in%20a%20near%20failure%20(or%20failure%20on%20nightly-testing)%20of%20the%20simpNF%20linter%20on%20%60Complex.comap_exp_cobounded%60%3A%0A%0A%2F--%0Aerror%3A%20failed%20to%20synthesize%0A%20%20AddMonoidHomClass%20(AddGroupSeminorm%20%E2%84%82)%20%E2%84%82%20%E2%84%9D%0A(deterministic)%20timeout%20at%20'typeclass'%2C%20maximum%20number%20of%20heartbeats%20(19000)%20has%20been%20reached%20(use%20'set_option%20synthInstance.maxHeartbeats%20%3Cnum%3E'%20to%20set%20the%20limit)%0A-%2F%0A%23guard_msgs%20in%0Aset_option%20synthInstance.maxHeartbeats%2019000%20in%0Aexample%20%3A%20comap%20exp%20(cobounded%20%E2%84%82)%20%3D%20comap%20re%20atTop%20%3A%3D%20by%20simp%0A%0A)

```
import Mathlib

open Complex Filter Bornology

/--
error: failed to synthesize
  AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ
(deterministic) timeout at 'typeclass', maximum number of heartbeats (19000) has been reached (use 'set_option synthInstance.maxHeartbeats <num>' to set the limit)
-/
#guard_msgs in
set_option synthInstance.maxHeartbeats 19000 in
#synth AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ

-- This then results in a near failure (or failure on nightly-testing) of the simpNF linter on
-- `Complex.comap_exp_cobounded` and `Complex.map_exp_comap_re_atTop`:

/--
error: failed to synthesize
  AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ
(deterministic) timeout at 'typeclass', maximum number of heartbeats (19000) has been reached (use 'set_option synthInstance.maxHeartbeats <num>' to set the limit)
-/
#guard_msgs in
set_option synthInstance.maxHeartbeats 19000 in
example : comap exp (cobounded ℂ) = comap re atTop := by simp
```

[Output](https://gist.githubusercontent.com/semorrison/47b4723c30a95f62762eb058ff85ec57/raw/2b41f5a8a4b97a09f28babbad6d20682d34d56b9/gistfile1.txt) of
```
set_option trace.Meta.synthInstance true in
#synth AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ 
```

**Comment by mattrobball** [comment]
Also fixed by leanprover/lean4#4085

**Comment by jcommelin** [comment]
I confirm that this is fixed:
```lean
import Mathlib

/-!
# Test against slow typeclass synthesis

This file tests against regression in typeclass synthesis speed.

See https://github.com/leanprover-community/mathlib4/issues/12226
for the initial issue, that was fixed by
https://github.com/leanprover/lean4/pull/4085
-/

open Complex Filter Bornology

/--
error: failed to synthesize
  AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ
-/
#guard_msgs in
set_option synthInstance.maxHeartbeats 3000 in
#synth AddMonoidHomClass (AddGroupSeminorm ℂ) ℂ ℝ

-- This then results in a near failure (or failure on nightly-testing) of the simpNF linter on
-- `Complex.comap_exp_cobounded` and `Complex.map_exp_comap_re_atTop`:

set_option synthInstance.maxHeartbeats 3000 in
example : comap exp (cobounded ℂ) = comap re atTop := by simp
```
